rpc/src/responses/json_array.rs

```diff
@@ -1,47 +1,280 @@
-    flattened_vec: Vec<T>,
+    items: Vec<T>,
 }
 
 impl<T: de::DeserializeOwned> JsonArray<T> {
     pub fn from_response_str(str_to_parse: &str) -> Result<Self, ParseError> {
-        let mut parse_response = serde_json::from_str::<Value>(str_to_parse)?;
-        let parsed_json_array = parse_response
+        let items = try_parse_array_into_item_from_response_str(str_to_parse)?;
+        Ok(Self { items })
+    }
+
+    pub fn into_vec(self) -> Vec<T> {
+        self.items
+    }
+
+    pub fn from_json_array(json_value: &mut Value) -> Result<Self, ParseError> {
+        let array = json_value
             .as_array_mut()
             .ok_or_else(|| generate_none_error("cannot parse initial json string as array"))?;
 
-        let mut flattened_vec = Vec::new();
-        for nested_entry in parsed_json_array {
-            let parsed_item_value = match nested_entry.is_array() {
-                true => unwrap_item_in_nested_json_array(nested_entry.take())?,
-                false => nested_entry.take(),
-            };
-            let converted_item: T = serde_json::from_value(parsed_item_value)?;
-            flattened_vec.push(converted_item);
+        let mut items = Vec::new();
+        for item in array {
+            let converted_item = serde_json::from_value(item.take())?;
+            items.push(converted_item);
         }
+        Ok(Self { items })
+    }
 
-        Ok(Self { flattened_vec })
+    pub fn from_json_object(json_value: &mut Value) -> Result<Self, ParseError> {
+        let converted_object = serde_json::from_value(json_value.take())?;
+        let items = vec![converted_object];
+        Ok(Self { items })
     }
 
-    pub fn into_vec(self) -> Vec<T> {
-        self.flattened_vec
+    pub fn len(&self) -> usize {
+        self.items.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.items.is_empty()
     }
 }
 
-fn unwrap_item_in_nested_json_array(mut nested_item: Value) -> Result<Value, ParseError> {
-    let nested_array = nested_item
-        .as_array_mut()
-        .ok_or_else(|| generate_none_error("invalid initial json array"))?;
-    match nested_array.last() {
-        Some(_) => Ok(nested_array.pop().unwrap()),
-        None => Ok(json!("")),
+impl<T: de::DeserializeOwned> JsonArray<JsonArray<T>> {
+    pub fn from_nested_response_str(
+        nested_json_str: &str,
+    ) -> Result<JsonArray<JsonArray<T>>, ParseError> {
+        let outer_array = try_json_str_into_json_array(nested_json_str)?;
+
+        let mut items = Vec::new();
+        for mut item in outer_array {
+            let json_item = match &item {
+                Value::Array(_) => JsonArray::from_json_array(&mut item),
+                Value::Object(_) => JsonArray::from_json_object(&mut item),
+                _ => {
+                    let detail = "initial json value is neither an object or an array".to_string();
+                    Err(ParseError::ResponseParsingError(detail))
+                }
```

Comment:
`serde_json::Value` is actually a really nice Enum that allows for some cute `match` statements